elk/training/eigen_reporter.py

```diff
@@ -1,35 +1,59 @@
 """An ELK reporter network."""
 
 from ..math_util import cov_mean_fused
-from .reporter import Reporter, ReporterConfig
+from .reporter import EvalResult, Reporter, ReporterConfig
+from copy import deepcopy
 from dataclasses import dataclass
-from torch import Tensor
-from typing import Optional
+from torch import nn, optim, Tensor
+from typing import Optional, Sequence
 import torch
-import torch.nn as nn
 
 
 @dataclass
 class EigenReporterConfig(ReporterConfig):
+    """Configuration for an EigenReporter.
+
+    Args:
+        inv_weight: The weight of the invariance term in the loss.
+        neg_cov_weight: The weight of the negative covariance term in the loss.
+        num_heads: The number of reporter heads to fit. In other words, the number
+            of eigenvectors to compute from the VINC matrix.
+    """
+
     inv_weight: float = 5.0
     neg_cov_weight: float = 5.0
 
+    num_heads: int = 1
+
 
 class EigenReporter(Reporter):
     """A linear reporter whose weights are computed via eigendecomposition.
 
     Args:
         in_features: The number of input features.
         cfg: The reporter configuration.
+
+    Attributes:
+        config: The reporter configuration.
+        contrastive_M2: The running sum of the cross-covariance between the
+            centroids of the positive and negative clusters.
+        intracluster_cov: The running sum of the covariance matrices of the
+            centroids of the positive and negative clusters.
+        M2: The running sum of the covariance matrices of the centroids of the
+            positive and negative clusters.
+        n: The running sum of the number of samples in the positive and negative
+            clusters.
+        weight: The reporter weight matrix. Guaranteed to always be orthogonal, and
+            the columns are sorted in descending order of eigenvalue magnitude.
     """
 
     config: EigenReporterConfig
 
-    n: Tensor
-
     contrastive_M2: Tensor
+    intracluster_cov: Tensor
     M2: Tensor
```

Comment:
The naming of these tensors is somewhat confusing (to me). It seems like they should be named something like `variance_M2`, `invariance_M2`, `neg_covariance_M2`, or `intercluster_cov`, `intracluster_cov`, `intracluster_xcov`. Also the docstring for M2 and intracluster_cov are the same.

Reply:
Can you commit this